Package name containing reserved Java keyword is not cleaned up

bono007 opened the pull request:
This PR adds Java reserved keywords to package name cleaning and fixes gh-983.

### Verification

* Added thorough parameterized unit tests

* Ran curl against locally running `initializr-service-sample` using custom package name with reserved keyword - verified it uses default name instead:

```
curl localhost:8080/starter.zip -o demo.zip -d packageName=com.import.bar
```
which contains...
```
IT-USA-10850:temp cbono$ unzip demo.zip
Archive:  demo.zip
  inflating: mvnw.cmd
  inflating: pom.xml
  inflating: .gitignore
   creating: .mvn/
   creating: .mvn/wrapper/
  inflating: .mvn/wrapper/MavenWrapperDownloader.java
  inflating: .mvn/wrapper/maven-wrapper.properties
  inflating: .mvn/wrapper/maven-wrapper.jar
  inflating: mvnw
   creating: src/
   creating: src/test/
   creating: src/test/java/
   creating: src/test/java/org/
   creating: src/test/java/org/acme/
   creating: src/test/java/org/acme/demo/
  inflating: src/test/java/org/acme/demo/DemoApplicationTests.java
   creating: src/main/
   creating: src/main/resources/
  inflating: src/main/resources/application.properties
   creating: src/main/java/
   creating: src/main/java/org/
   creating: src/main/java/org/acme/
   creating: src/main/java/org/acme/demo/
  inflating: src/main/java/org/acme/demo/DemoApplication.java
```

pivotal-issuemaster commented:
@bono007 Please sign the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=spring-io/initializr&pullRequestId=1018)!

[Click here](https://cla.pivotal.io/sync/pivotal?repositoryId=spring-io/initializr&pullRequestId=1018) to manually synchronize the status of this Pull Request.

See the [FAQ](https://cla.pivotal.io/about) for frequently asked questions.

pivotal-issuemaster commented:
@bono007 Thank you for signing the [Contributor License Agreement](https://cla.pivotal.io/sign/pivotal?repositoryId=spring-io/initializr&pullRequestId=1018)!

bono007 commented on CONTRIBUTING.adoc:
```diff
@@ -26,7 +26,7 @@ given the ability to merge pull requests.
 added after the original pull request but before a merge.
 
-* Use the https://github.com/spring-projects/spring-framework/wiki/Spring-Framework-Code-Style[Spring Framework code format conventions].
+* Use the https://github.com/spring-projects/spring-framework/blob/master/CONTRIBUTING.md#source-code-style[Spring Framework code format conventions].
```
[Unrelated to issue 983] I noticed that the original link did not go anywhere near the source code style page. However, the direct link actually references the `CONTRIBUTING.md` file in master path. I am happy to revert this but thought it may be helpful.

bono007 commented on initializr-metadata/src/main/java/io/spring/initializr/metadata/InitializrConfiguration.java:
```diff
@@ -155,6 +161,10 @@ private static boolean hasInvalidChar(String text) {
 	}
 
+	private static boolean hasReservedKeyword(final String packageName) {
+		return Arrays.stream(packageName.split("\\.")).anyMatch(SourceVersion::isKeyword);
```
I know of the improvements made in Boot 2.2 by **not** using streams, but I doubt this is in a tight loop where it will have much impact.

snicoll replied:
It's fine.

bono007 commented:
@snicoll I rebased onto `upstream/master` and removed my mod to the `CONTRIBUTING.md`. I was about 95% sure that it should not have been in this commit. I will keep that in mind going forward. Onto the next ticket...

snicoll commented:
@bono007 thank you very much for making your first contribution to Spring Initializr. I've [polished the tests](https://github.com/spring-io/initializr/commit/9143e96b01802a2f8a6ec95d1060cb059648bd31) as I felt it was more testing the contract of `SourceVersion` than your change.

bono007 commented:
@snicoll the reason I wrote the test w/ the keywords checks is to not take advantage of the fact that we are leveraging a well tested SourceVersion class. If we swapped it out w/ an internal imp I would still want the test to verify the keywords.  Also, w/ Junit5 it was inexpensive to do so. 

That being said, unit testing in general does take advantage of the current implementation (aka other units) and expects that those other "units" are also well tested.

snicoll commented:
>  If we swapped it out w/ an internal imp I would still want the test to verify the keywords.

A counter argument to this is that if `SourceVersion` is improved in the future with an additional keyword, there's nothing that will auto-update the list you've copied in the test. To me it feels odd that we test a very specific set of keywords that may become out-of-date with a new Java version.

Anyway, We agree we shouldn't be testing that `SourceVersion` does what it's supposed to do here anyway :)
